pgmpy/datasets/datasets.py

```diff
@@ -0,0 +1,342 @@
+
+                    if u in G and v in G:
+                        G.add_edge(u, v)
+                except Exception:
+                    continue
+        return DAG(G)
+
+    except Exception:
+        return None
+
+
+def parse_tier_graph(
+    text_content: str,
+) -> Optional[DAG]:  # Parses the 'abalone.knowledge.txt' tiered format.
+    try:
+        lines = [line.strip() for line in text_content.splitlines() if line.strip()]
+
+        start_index = -1
+        for i, line in enumerate(lines):
+            if line.lower() == "addtemporal":
+                start_index = i
+                break
+
+        if start_index == -1:
+            return None
+
+        G = nx.DiGraph()
+        tiers = []
+
+        for line in lines[start_index + 1 :]:
+            match = re.match(r"^\d+\s+(.*)", line)
+            if match:
+                vars_list = match.group(1).split()
+                if vars_list:
+                    tiers.append(vars_list)
+                    G.add_nodes_from(vars_list)
+
+            elif line.startswith("/") or "direct" in line.lower():
+                break
+        for i in range(len(tiers)):
+            for j in range(i + 1, len(tiers)):
+                for u in tiers[i]:
+                    for v in tiers[j]:
+                        if u in G and v in G and u != v:
+                            G.add_edge(u, v)
+
+        return DAG(G)
+
+    except Exception:
+        return None
+
+
+def parse_knowledge_txt(raw: bytes) -> Optional[DAG]:
+    text = raw.decode("utf-8-sig", errors="ignore")
+
+    sachs_graph = parse_sachs_graph(text)
+    if sachs_graph is not None:
+        return sachs_graph
+
+    tier_graph = parse_tier_graph(text)
+    if tier_graph is not None:
+        return tier_graph
+    return None
+
+
+class BaseDataset:
+    name: str = ""
+    description: str = ""
+    tags: Dict[str, Any] = {}
+
+    @classmethod
+    def cache_path(cls, key: str) -> str:
+        ensure_dir(PGMPY_DATA_HOME)
+        safe = hashlib.sha256(f"{cls.name}:{key}".encode()).hexdigest()[:16]
+        ext = ".bin"
+        if key.startswith("data"):
+            ext = ".parquet"
+        elif key.startswith("ground_truth"):
+            ext = ".json"
+        return os.path.join(
+            PGMPY_DATA_HOME, f"{cls.name}_{key.replace(':','-')}_{safe}{ext}"
+        )
+
+    @classmethod
+    def _fetch(cls, key: str) -> bytes:
+        raise NotImplementedError
+
+    @classmethod
+    def load_or_fetch(cls, key: str):
+        cache_path = cls.cache_path(key)
+        if os.path.exists(cache_path):
+            with open(cache_path, "rb") as f:
+                raw = f.read()
+        else:
+            raw = cls._fetch(key)
+            with open(cache_path, "wb") as f:
+                f.write(raw)
+        return raw
+
+    @classmethod
+    def load(cls):
+        raise NotImplementedError
+
+    @classmethod
+    def load_ground_truth(cls):
+        return None
+
+
+def load_dataset(
```

Comment:
I think we might need to add a `variant` argument here to allow users to choose which variant to load. It can be `None` by default which loads the default variant of the dataset.